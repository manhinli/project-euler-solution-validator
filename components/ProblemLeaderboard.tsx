import React from "react";
import useSWR from "swr";
import styles from "./ProblemLeaderboard.module.css";
import { ProblemLeaderboardEntry } from "../types/Problem";

// Client-side local region-dependent formatting of date/time
const SOLUTION_DATETIME_FORMAT = Intl.DateTimeFormat(undefined, {
    dateStyle: "short",
    timeStyle: "medium",
});

export interface Props {
    /** ID of the problem */
    problemId: string | number | undefined;
}

export const ProblemLeaderboard: React.FC<Props> = ({ problemId }) => {
    const { data, error } = useSWR<readonly ProblemLeaderboardEntry[]>(
        problemId ? `/api/problem/${problemId}/leaderboard` : null,
        // Update leaderboard every minute
        { refreshInterval: 60000 }
    );

    if (error) {
        return (
            <div className={styles.error}>
                Leaderboard could not be retrieved.
            </div>
        );
    }

    return (
        <table className={styles.leaderboard}>
            <thead>
                <tr>
                    <th>Name</th>
                    <th>Submission time</th>
                    <th>Number of attempts</th>
                </tr>
            </thead>
            <tbody data-testid="table-body">
                {data?.map(
                    ({
                        userName,
                        earliestSuccessfulAttempt,
                        numberOfAttempts,
                    }) => (
                        <tr key={userName}>
                            <td>{userName}</td>
                            <td>
                                <time
                                    dateTime={earliestSuccessfulAttempt}
                                    title={earliestSuccessfulAttempt}
                                >
                                    {SOLUTION_DATETIME_FORMAT.format(
                                        new Date(earliestSuccessfulAttempt)
                                    )}
                                </time>
                            </td>
                            <td>{numberOfAttempts}</td>
                        </tr>
                    )
                )}
            </tbody>
        </table>
    );
};

export default ProblemLeaderboard;
