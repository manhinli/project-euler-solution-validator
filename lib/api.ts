import { NextApiResponse } from "next";

type Json = Record<string, unknown> | Json[];

/**
 * Sends HTTP 200 OK back to client with given data.
 *
 * @param res Next.js response object
 * @param data Data to send to client
 */
export function respondSuccessWithJson(res: NextApiResponse, data: Json) {
    return res.status(200).json(data);
}

/**
 * Sends a client error response (defaulting to HTTP 400) with given error.
 *
 * @param res Next.js response object
 * @param code Response code
 * @param error Error object
 */
export function respondClientError(
    res: NextApiResponse,
    code: number = 400,
    error?: unknown
) {
    let message = "Bad request";

    // Determine if `error` has a message that we can provide
    if (error && typeof error === "object" && Object.hasOwn(error, "message")) {
        message = `${(error as any).message}`;
    } else if (typeof error === "string") {
        message = error;
    }

    return res.status(code).json({
        error: message,
    });
}

/**
 * Sends HTTP 404 Not Found back to client.
 *
 * @param res Next.js response object
 */
export function respondNotFound(res: NextApiResponse) {
    return respondClientError(res, 404, "Not found");
}

/**
 * Sends HTTP 405 Method Not Allowed back to client.
 *
 * @param res Next.js response object
 */
export function respondMethodNotAllowed(res: NextApiResponse) {
    return respondClientError(res, 405, "Method not allowed");
}
