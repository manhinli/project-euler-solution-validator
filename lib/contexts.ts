import React, { Dispatch, SetStateAction } from "react";

export const UsernameContext = React.createContext<{
    value: string;
    onChange: Dispatch<SetStateAction<string>>;
}>({
    value: "",
    onChange: () => {},
});
