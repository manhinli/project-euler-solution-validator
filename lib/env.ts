export const ENVIRONMENT_VARIABLES = [
    "DATABASE_URI",
    "DATABASE_NAME",
    "PROBLEMS_COMPILED_PATH",
] as const;

export function getEnv(variable: typeof ENVIRONMENT_VARIABLES[number]) {
    const value = process.env[variable];

    if (!value) {
        throw new Error(`${variable} not defined`);
    }

    return value;
}
