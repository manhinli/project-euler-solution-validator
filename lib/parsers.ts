import { z } from "zod";

/**
 * Parses a value that is intended to be a problem ID.
 *
 * @param id Value to be parsed as problem ID
 * @returns {number} Problem ID value
 */
export function parseProblemId(id: any): number {
    const schema = z
        .number()
        .int("Problem ID must be integer")
        .positive("Problem ID must be positive integer");

    return (
        z
            // Apply value transforms before testing against schema
            .preprocess((value) => {
                // If already number, continue
                if (typeof value === "number") {
                    return value;
                }

                // If string, parse as number
                if (typeof value === "string") {
                    return Number.parseFloat(value);
                }

                // Otherwise ignore value
                return undefined;
            }, schema)
            // Parse value
            .parse(id)
    );
}
