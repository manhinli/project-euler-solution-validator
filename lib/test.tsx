import "@testing-library/jest-dom";
import { SWRConfig } from "swr";

// Provide a fresh provider every time we use SWR
export const swrConfigWrapper = (
    children: any,
    config: React.ComponentProps<typeof SWRConfig>["value"] = {}
) => (
    <SWRConfig value={{ provider: () => new Map(), ...config }}>
        {children}
    </SWRConfig>
);

// Utility to deep clone objects
export const deepClone = (o: object) => JSON.parse(JSON.stringify(o));
