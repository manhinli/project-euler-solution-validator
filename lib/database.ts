import { Collection, MongoClient } from "mongodb";
import { getEnv } from "./env";

export const DATABASE_URI = getEnv("DATABASE_URI");
export const DATABASE_NAME = getEnv("DATABASE_NAME");
export const COLLECTIONS = ["attempts", "problems"] as const;

// Singleton MongoDB client instance
let mongoClient: MongoClient | undefined = undefined;

export async function getClient() {
    if (mongoClient) {
        return mongoClient;
    }

    const client = new MongoClient(DATABASE_URI);
    mongoClient = await client.connect();
    return mongoClient;
}

export async function getDatabase() {
    const client = await getClient();
    return client.db(DATABASE_NAME);
}

export async function getCollection<T>(
    collection: typeof COLLECTIONS[number]
): Promise<Collection<T>> {
    const db = await getDatabase();
    return db.collection(collection);
}
