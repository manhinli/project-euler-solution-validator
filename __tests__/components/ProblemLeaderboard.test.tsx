/**
 * @jest-environment jest-environment-jsdom
 */

import { getAllByRole, render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import ProblemLeaderboard from "../../components/ProblemLeaderboard";
import { swrConfigWrapper } from "../../lib/test";

describe("ProblemLeaderboard", () => {
    it("renders expected table header", () => {
        render(<ProblemLeaderboard problemId={42} />);

        const columnHeaders = screen
            .getAllByRole("columnheader")
            .map((x) => x.textContent);

        expect(columnHeaders).toEqual([
            "Name",
            "Submission time",
            "Number of attempts",
        ]);
    });

    it("renders data from API", async () => {
        const data = [
            {
                userName: "First User",
                earliestSuccessfulAttempt: "2020-06-30T23:59:59.999Z",
                numberOfAttempts: 16,
            },
            {
                userName: "Second User",
                earliestSuccessfulAttempt: "2022-03-26T08:26:31.582Z",
                numberOfAttempts: 32,
            },
        ];

        const fetcher = jest.fn(() => Promise.resolve(data));

        render(
            swrConfigWrapper(<ProblemLeaderboard problemId={42} />, { fetcher })
        );

        const getTableBodyDataRows = () =>
            getAllByRole(screen.getByTestId("table-body"), "row");

        await waitFor(() => getTableBodyDataRows().length > 0);

        expect(getTableBodyDataRows()).toMatchSnapshot();
    });

    it("renders error message when service unavailable", async () => {
        const fetcher = jest.fn(() => {
            throw new Error();
        });

        render(
            swrConfigWrapper(<ProblemLeaderboard problemId={42} />, { fetcher })
        );

        const getErrorMessage = () =>
            screen.getByText("Leaderboard could not be retrieved.");

        await waitFor(() => getErrorMessage());

        expect(getErrorMessage()).toBeInTheDocument;
    });

    it("does not fetch when problem ID not provided", async () => {
        const fetcher = jest.fn(() => Promise.resolve());

        render(
            swrConfigWrapper(<ProblemLeaderboard problemId={undefined} />, {
                fetcher,
            })
        );

        expect(fetcher).not.toBeCalled();
    });
});
