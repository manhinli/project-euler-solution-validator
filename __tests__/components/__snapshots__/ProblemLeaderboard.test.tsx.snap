// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProblemLeaderboard renders data from API 1`] = `
Array [
  <tr>
    <td>
      First User
    </td>
    <td>
      <time
        datetime="2020-06-30T23:59:59.999Z"
        title="2020-06-30T23:59:59.999Z"
      >
        6/30/20, 11:59:59 PM
      </time>
    </td>
    <td>
      16
    </td>
  </tr>,
  <tr>
    <td>
      Second User
    </td>
    <td>
      <time
        datetime="2022-03-26T08:26:31.582Z"
        title="2022-03-26T08:26:31.582Z"
      >
        3/26/22, 8:26:31 AM
      </time>
    </td>
    <td>
      16
    </td>
  </tr>,
]
`;
